### DISCUSSION TITLE
On-Line Creation (OLC) Story

### DISCUSSION THREAD

**Discussion by Irthwynd** [DavidRieman/WheelMUD#89] [General]
I took a good look at what you presented; I think it looks fantastic for, as you say, a genre-agnostic building tool. I held the positions of world builder and class designer for a game that exclusively used XML files for every room, object, NPC, etc. It was cumbersome, and with so many moving parts it was easy to miss a field or deposit the file in the wrong directory. However, it did have some advantages in that it allowed me to take out batches of files I could organize and work on in a very tactile way. All to say, my view might be biased because how I built was a bit more technical than most. From the perspective of your OLC, this feels like it has a much easier barrier for entry and would allow developers to jump into the action with very little training. I would hope everything enters a QC directory so senior staff or admin could proofread and give the green light, but that's just a little side note. The only thing I would *personally* want is a way to import these created files into notepad++ or some other editing program if needed, but overall I think this is very user friendly and much less tedious compared to my experience.

These are my initial impressions and I do have more specific suggestions, but I am not sure if they would break the genre-agnostic nature of your project. Would you prefer I created another discussion thread?

**Comment by DavidRieman** [comment]
Thanks for the feedback!  (This thread is referencing [this OLC User Story design doc](http://htmlpreview.github.io/?https://github.com/DavidRieman/WheelMUD/blob/master/docs/Design/UserStory_OnLineCreation.htm)).

> hope everything enters a QC directory

Yeah, the intent for the default WheelMUD setup - the tools we give "out of the box" - is that a high-level admin could have permissions to edit things directly, while entry-level builders could be given access only to editing Things that are within assigned "zones" (AKA "areas" to some MUDs).  A "zone" Thing would be the actual "Parent" of all things in that zone, with a behavior that would help protect against bad things from happening (such as denying any Player move request into any place within that zone - I'd like it to be safe by default to have builders working on the same game instance instead of requiring a separate "builder port" if possible).  The builder could mark the zone as ready for review with a command like "zone submit".  An admin or builder with sufficient permissions could see all pending submissions with another command like "zone review", including a summary of any outstanding validation warnings and such.  So the zone couldn't be attached to the rest of the player-inhabited part of the world until it was admin-approved.

An editorial mode would be neat for zone reviews - showing you one unreviewed Thing at a time, that you could mark reviewed with a single key and repeat that process until all Things are marked approved.  Or another key to enter direct editing to revise simple typos or grammar.  Or another key to leave a note on the pending Thing.  Basically like a GitHub code review comment attached to that Thing ID.  Imagine attaching a note to a mob "It's weird that this goblin is stronger than our dragons and that doesn't fit our theme of progression; can we find another mob type to use here?"  The builder could list those unresolved notes, address them, respond, and resubmit.  That'd give us a nice healthy feedback loop for zones to get iterated until ready for release.  Zone review systems would probably make for another nice User Story document.  :)

(aside: I'd also like validation rules to be easily extensible - so you could write your own code to check details specific to the Game System, like raising a warning if a mob feels "imbalanced" against expectations for their level.)

> a way to import these created files into notepad++ or some other editing program

Yeah, in the past we had at least two separate small efforts that I recall to have "offline world builder" tools built (for native Windows and a web interface in JavaScript IIRC) where the goal was to either be able to connect directly to our same DB directly, or provide import/export techniques - but the main ideas were to give the builder a GUI for world-building.  Neither of these efforts got very far and wouldn't be compatible with today's Document DB approach if they had anyway.  Another thought would be scripts that take the output format of _existing_ offline world builders meant for other MUD types and somehow import it to WheelMUD.  In short, I agree.  It's a topic we'll need to revisit later in the roadmap though (currently marked at v1.1+), after we get much more experience working with the Document DB and get a better idea what kind of options would be feasible.

**Reply by ghost** [reply]
It would be interesting to theorycraft around the idea of using GitHub itself as a core part of the content development pipeline. For example, want to alter a room? Fork the content repo and submit a PR to the appropriate directory (zone) and file (room.json).

**Reply by Irthwynd** [reply]
>Basically like a GitHub code review comment attached to that Thing ID. 

I was world building in the early/mid 2000s before GitHub was a thing. It never occurred to me it would be a good model for the developmental process, but by golly I agree.

Something I did not clearly convey: world building with xml was certainly a pain in the ass, but once I did have those 50+ room files for a region in tabs, I could easily cycle through to check for continuity, fine tune, or restructure when necessary as every detail was at a relatively quick glance. Who knows if others would find it as useful as I did. Just my two cents, really.

**Reply by DavidRieman** [reply]
I think integrating GitHub into a world-building process for a MUD on WheelMUD would involve an alternative to RavenDB for storage.  The framework is built to be extensible and configurable in this manner though.  For example, one could build a class on `IWheelMudDocumentStorageProvider` which saves and loads the Thing documents via regular out-of-the-box .NET XML Serialization into a clever file structure (like in a folder for whichever Zone the Thing is attributed to, and file named via the Unique ID).  Thus you could build your world with hand-editing through text editors or OLC against your locally-running dev server instance and push those changes through a GitHub PR processes.  There'd be _pros and cons_ to this approach of course, but it would be neat to try to have an XML Serialization Document Storage Provider available in Core WheelMUD for those who want to use it.

**Comment by DavidRieman** [comment]
Seems like the best place to put some OLC threads from the Archives, for further clarification and context.  :)

_This thread contains restored content from prior forums; see #134 for details._

[July 2010]

> Could you nest behaviors inside of behaviors?

Sure, sort of.  Say you have an OnOffSwitchBehavior attached to the interior of a tunnel that "fills quickly with water" when the switch is thrown.  Normally a room doesn't just suddenly gain "underwater" qualities on the fly but in this case you want that to happen, so your OnOffSwitchBehavior applies the UnderwaterBehavior to the room.  (At this point we definitely also want properties of type SensoryMessage on the OnOffSwitchBehavior like "OnCreatesMessage" and "OffCreatesMessage" - and the builder would apply an appropriate sensory message for the act of being suddenly surrounded in water or water draining away.)

> Do behaviors/effects only go two levels deep?

No, it is totally dynamic based on the properties of each behavior.  Most behaviors/effects will not have a reason to house other behaviors/effects, but will instead contain properties of only fairly basic types (int/string/SensoryMessage/etc.)

> How do you list all behaviors?  How do you list all effects?

Per this user story, it could be simply 'list behaviors' or 'list effects'.  Depending on your current telnet client and MUD settings, this may automatically cause paging pauses when there are a lot of things to list.

> each MUD to have a reference sheet of sorts?

I think having smart features with the list commands would be sufficient;  IE there could be ways to sort the results by the most popularly referenced behaviors and such first, which a builder could then be confident are well tested and likely to be complete, standardized, etc.

> Can you specify what you want the object id to be?  I like to be quite organized with my design and would want to use number ranges for similar types of objects.  Even better if it's alphanumeric.

Maybe.  I was currently thinking the ID would always match the DB-assigned ID.  I could see there being MUD administrative commands for allocating a 'block' of Thing IDs to a given builder/zone (by persisting a block of blank Things simultaneously with the appropriate builder permissions attached, or whatnot).  But I think instances of behaviors and such are probably best left dynamic, since you can't/shouldn't share the behavior instance itself.  If you're building two Swords, they each get their own individual WeaponBehavior instances, even if both their WeaponBehaviors are going to have similar properties, etc.

_update 2021_: These days I think we're more likely to see imposing "total things" or "total complexity" limits on the area being built now, that could be quickly be changed by admins (generally after looking at what's being built so far) - rather than needing "ID blocks" as were more familiar to certain old MUD frameworks of the past.

> Did that have a keyword of "light" or did it detect light in the object name?  Can you do wildcard searching?

Yes that's what I was intending to demonstrate.  Explicit wildcard system would be a good idea too although in lieu of any wildcards or whatnot being supplied I think a partial text match is probably ideal as the default.

Thanks for the feedback!

---

[Apr 2010]

> I like the idea of being able to say "new item" and then edit that object into a flashlight, sword, door, stargate, etc but I think we really need to make sure a class hierarchy exists, or even better: doesn't have to. It would be nice and flexible if we made the different types of items just instances of a general item that had attributes to allow them to do special things.
For example: You used a 'item' here.  Am I to understand that if the correct properties are set, this one 'item' could be a container, some furniture, etc?   That the goal is to not have to have a 'objmob' or 'objexit', etc, but one Thing that can be anything?  That would be best I think.

Yeah, we've been moving "lots of specifically coded item types" into the direction you're talking about (as seen in the Interfaces and Behaviors discussion).  For example, we no longer have a "Weapon.cs" - if you want to make a thing behave as a weapon, you give it WeaponBehavior and define the parameters.  If you want to make it edible, you'd give it like an EdibleBehavior.  Want a fun one-off case where you can wield your stale bread like a bat and use it to hit things for comically low damage?  No longer have to write copy-pasta code, you just attach _both_ an EditableBehavior and a WeaponBehavior to your Thing, and away you go.

CreatePortal.cs is an interim test command that demonstrates creating a Thing and attaching a Behavior to it that has interesting capabilities, on the fly.

> I was thinking about the consumable object story and how this is connected and it seems to me like everything is a consumable object.  Wood items of any type should be able to burn into ash, metal objects of any type should be able to melt into slag if enough heat is used and the object is not protected, crystal should shatter if its weak enough, etc.

I think a game that wants to go this direction, would start building related Behaviors into their prefab templates to force their builders to define these things.  One game might want all weapons to consider whether they are burnable/meltable while another team may feel that's not a direction they want for their game.  Hopefully even if a game starts bolting such systems on after Things were already designed, as long as things had a good "based on" chain marked up - they'd get notified what items need to be updated when they added such Behaviors to the base prefabs/templates...  Assisting immensely with the process of upgrading all existing objects to the new expectations.

> As far as flexibility for object creation goes, from my perspective:
> 1. Containers are just doors to very small and very private rooms that most - note I did not say all - players will not be able to fit into. 
 They move with the player, can be locked from the outside or the inside, can be punctured or damaged (and thus destroyed)
> 2. Even a stack of paper can attack you and give you a very brutal paper cut, or move from room to room like a traveling 'mob' if there was a nice breeze blowing it around.
> 3. Think about a wooden chair. It can be an item, a place the player can sit, a consumable, a projectile if you throw it.  It can be picked up, thrown against the ground or another player, do damage to what it hits, or itself if it lands wrong.  If it is destroyed the parts that are left could be used in other ways: A chair leg used as a club, or stake to use against vampires, or the wood could simply be burned to create light or heat in the form of a torch.
> It would be nice if our OLC story supports this, without having to create a bunch of sub-type objects. Every object has a consumable type as a 'this is the base type' type.  A sword is metal, a torch is mostly wood, a chair is mostly wood in the example above, and I can sit on a rug or a sword - ouch - just as easily as I could sit on a chair.</div>

Lots of unique ideas there.  Indeed I feel the core, compositional "Things have Behaviors" technical design supports such things as possible game systems.

---

[July 2010]
> Where I thought it was a bit confusing is where you started with a flashlight, then went into editing the behavior, then the effect.  It's not clear that they are all tied together.  Instead of simply saying "You are editing LightingEffectID45", I would rather see "You are editing LightningEffectID45 of OnOffSwitchBehaviourID34 of ObjectID123 (a flashlight)."  Or something like that.  Or maybe a command that shows what you are currently working on and how it links together.  Or maybe change the new behavior/effect command to something more like "attach" to object.

Sure, chaining, or even some sort of tab indentation formatting, may be helpful.  We could try to design or experiment with something there.  I think, at least for MXP-supported clients, we could even go heavy on hyperlinking.  So any time you see some ID referenced, you can click it to view _that_ object and drill in naturally!

> Can you specify what you want the object id to be?  I like to be quite organized with my design and would want to use number ranges for similar types of objects.  Even better if it's alphanumeric.

Haven't really designed any kind of non-dynamic ID capabilities yet - but it's a good idea to have such an option eventually.  As long as it doesn't make the auto-ID system complicated.  Maybe we could do something like allow "namespaces" for Things, and can assign a Builder access to use a given namespace.  So a builder named Bob could be given the "bob" namespace, and then specify their Thing's ID to be "bob/1" but it would still be attached to the current Zone, etc.  Our document storage system(s) should handle such IDs fine.  We'll want to think on this again after OLC starts to get up and running.

> Also I'd be worried that I would create duplicate objects because they hadn't had the right keywords set.  Different builders would use different keywords for the same object.  For example, in the UK a flashlight is called a torch.  In the U.S. a torch is a burning stick.  

Yeah word meanings and spellings are going to be an issue with any international game, MUD or otherwise.

> It looked like the keywords were not case sensitive and had some sort of fuzzy logic?  You searched for "light" and "LightningEffect" came up.  Did that have a keyword of "light" or did it detect light in the object name?  Can you do wildcard searching?

No fuzzy logic yet - "light" is a substring of "lightning" so a simple case-insensitive "contains" search will turn up such a hit.  Potentially we could build in tech to handle nearby search relevance and typos and such.  _Eventually_.  Not sure I'd want such complexities going in the core code base until the v1.1+ time frame, and hopefully it'll be something that can be "plugged in" to the core command targeting systems (not just OLC).  So it could be considered by each admin for trade-offs in computing power and relevance and so on, as there are several algorithms one could choose from.

---

[July 2012]

In (the original version of) this user story I showed how to create a new object _based on_ an existing object:

$ new object from 123
Object created with ID130 and properties inherited from a flashlight.

It occurs to me that some builders may wish to provide other builders with preconstructed templates _specifically designed for_ new objects to be based on.  For a game expecting there to be a lot of "mounts" created, the experience might set up to be like this:

$ new mount
Thing ID199 created with properties inherited from **mount**.

$ prop deep
... shows that it was created with a full set of behaviors like MobileBehavior, MovableBehavior, LivingBehavior, MountBehavior, and so on as a senior builder deemed appropriate as defaults for a mount.
One might call this <i>prefabricating</i> and thus define any Thing so prepared to be <i>a prefab</i>.  (This is similar to what the video games industry typically uses the term for, IE you might prefabricate a boulder or tree and instance it in many places through the world with a few changing details, like scale and orientation.)
This could simplify the common builder workflow, as they could list prefabs and see "Room, Door, Hall, Vehicle, Mount, Mob, Weapon, Armor..." and know they're making an object the game should recognize as a weapon when they then use "new weapon" - and for senior builders, they're essentially guiding what they consider Best Practices for Building by providing a good set of prefabs for their world.

Basically I think we can support "prefabs" with some pretty simple mechanisms.  A senior builder would be able to simply tag a Thing as a prefab, giving it a unique prefab name (like "mount" in the example).  Storage-wise, our list of prefabs is just a pairing of **UniqueName,ID**.  The OLC "list objects ..." command could also list matching prefabs and call them out (on top, maybe with a different color or heading).  There could be a "list prefabs" command to filter to listing them specifically.  The OLC "new" command would look for matching prefab names when it didn't otherwise recognize the first/only argument; if found it would use the linked ID to base the new object off of.

(Note the latest OLC User Story hosted tackles much of this thinking.)

**Comment by DavidRieman** [comment]
# In-game World Building (OLC)

_This thread contains restored content from prior forums; see #134 for details._

[Feb 2010] This thread is my original proposal for OLC design. While this covers some topics not covered elsewhere yet, note that [the OLC User Story design doc](http://htmlpreview.github.io/?https://github.com/DavidRieman/WheelMUD/blob/master/docs/Design/UserStory_OnLineCreation.htm) is more accurate to our current vision for the parts that it does cover.

## Overview
This details a design for an **in-game** world building system which reduces a lot of the typical overhead/friction that normally prevents players from getting started with world building.

## Builder Characters
A player named “Bob” could instead chose to login to the same server with character name “builder:Bob” (using their normal password). This will log in with a basic “builder-flagged character” that has the name of Bob. This character has no equipment, can not leave the builder zones, and logs into the basic “Builder Hub” room. They can still OOC chat and do tells and such as normal, but other than that should not be able to go interact with other players or bring players to the builder zone, etc. The builder flag gives them access to a Builder chat channel. They can immediately create a new zone, which is tied to their player ID. A builder can not enter any other builder’s zone unless they have been explicitly granted permission to do so by an owner of the zone. (Administrators of sufficient privileges will always have access to all zones.) A zone-owner can use a command to grant another builder access to their zone, or even to grant co-ownership of the zone as an all-inclusive means to allow collaboration between builders. By default, a builder is only allowed to be owner/co-owner of maybe a couple zones at a time. Each zone could have budgeted limits for allowed number of rooms, items, mobs, and such as well.

### Builder Commands
Commands can be used to list the area(s) you own, and to list the rooms you’ve created, mobs you’ve created, etc. ‘Goto’ is only allowed for rooms tied to a zone you have builder access to, or allowed if you have administrative privileges. Ideally, rather than having dozens of specialized commands like ‘set-room-title’ we should have a couple flexible commands which set various attributes of given objects. For instance ‘setroom title The Buttery’ would set the title while ‘setroom desc’ would place you in the multi-line editor mode for writing the room’s description. (This multi-line editor mode would be another reuse of the edit mode underlying the ‘note system’ as discussed elsewhere.) 

#### Validation Command
An area verification command will help make the builders’ lives easier.
Some of these things we may be able to automatically enforce, but where we can’t, one verify command will list a bunch (or all) ‘warnings’ applicable to your zone, rooms, room elements, mobs, or items as specified. Verification steps which are executed by the verify command could be pulled together from the core engine and from the game system through MEF. Each verification step would perform its own iteration of all the zone’s rooms or mobs or items or whatnot, and simply generate warnings. Some example verification steps would be to generate warnings against:
* Any items in the zone that do not spawn anywhere, nor on any mob.
* Any item whose one-line description does not include appropriate color highlighting for an applicable keyword (or at least if not auto-enforcable).
* Any mobs that do not spawn anywhere (nor have a program to be spawned).
* Any mobs whose one-line description does not include appropriate color highlighting for an applicable keyword (or at least if not auto-enforcable).
* Any room examine elements (like for ‘examine pool’ where you didn’t create a ‘pool’ object) which have no keyword that is appropriately color-highlighted in the room’s description.
* Any room missing exits, or better yet, any room which can not be pathed to from the starting room (taking portals and the like into consideration wherever reasonable to do so).
* Use of disallowed color codes in any description element.
* Use of any non-alphanumeric special characters such as ‘beep’ characters.
* Any item in the zone whose power rating is outside acceptable bounds; this would be implemented in the game-system layer instead of the core layer.

## Tester Characters
Similar to the builder characters, a player named “Bob” could chose to login with a character name “tester:Bob” to log in with a basic “tester-flagged character” that has the name of Bob. This logs in as a plain character _in the core builder room_, with similar movement restrictions that builders have. Tester characters have access to no area-altering commands, but they should be provided commands from the game-system-layer for quickly altering the test character’s level and class, setting skills, stats, equipment and such to reasonable expectations for a run-of-the-mill human character of that level and class. An area can be marked as "ready for testing" by the builder(s) of the area, thus allowing any tester character to enter the zone. The intent is to provide a means of testing complicated content before going live with it. For example, wildly guessing at appropriate stats and abilities for a special-MUD-wide-event encounter with an uber-dragon is one thing, but having a few testers pit against it to see what progress they can make before getting slaughtered is better empirical data to tweak from.

## Publishing Content
An administrator with area-publishing privileges will first run the ‘verify’ command against the zone. Warnings should either be cleaned up by a builder or receive an exception by an administrator before the zone can become accepted into the official world. The overall zone info itself should be taggable with notes that are only visible to administrators (just as characters and such should be taggable with such notes). An administrator could tag the zone with their exception notes if all verification warnings have been reasoned to be acceptable. The administrator could then mark the zone as ‘testable’ to make it available for testing, or ‘published’ to make it available to players and strip the zone of all builder privileges so it can no longer be modified by a non-administrator. Some ‘glue’ would then need to be added to make the zone actually accessible to players. In many cases, this may be as simple as adding an exit to an existing production room which connects to the starting room of the newly-published zone, or adding a Portal that spawns somewhere in a production room that takes the players to the starting room, etc.

## Importing Content
Offline/desktop world editors, such as the Universe Manager Lite, can be used to create zones more efficiently than one can through the telnet protocol alone. These tools should be able to generate a single file that contains all needed zone information, which can then be sent to an administrator, who can then import the zone as a new zone in the game’s database. This imported zone will start out in the ‘being built’ state, preferably with the original builder’s character attached as the area-owner. The same in-game verification, testing, iteration, and publishing processes could then be used as the in-game process. If the builder wishes to iterate and provide a new version through their editor tool, the administrator could simply use an area-nuke command to delete the existing ‘being built’ zone and then import the new version.

## Challenges and Concerns
Using a single server to host both the game play and world-building activities has several advantages over the traditional approach of hosting two servers with one for play, and one for world building:
* Stronger community. You are not separated from everyone while you are world-building; you can still communicate with them, see them online, etc. The ‘who’ list better reflects everyone who is currently engaged with the game, and in what capacities.
* Better access to administrators on both sides. Administrators are a limited commodity who can not necessarily be effectively interacting with multiple world servers at the same time, but can better prioritize their interactions with a single world server.

However there are some notable disadvantages that need to be considered:
* The world-building experience has to be made quite stable since crashes, slowdown, and the like, also affect the actual players.
* Related to the above point, the attack area for malicious users is increased. To mitigate this, builder characters have to be _prevented_ as best we can from doing things like building weapons that cause randomization of trillions of dice per single attack, then getting said weapon to spawn and activate during ‘testing’, essentially bringing the server to a crawl. They cannot be allowed to create custom scripts that can spawn infinite things, or crash the game directly, etc. However, if we do not allow such ‘scripts’ and such to be built through the data only, it is obvious that, just like any other _code changes_, yes this new content has to be tested with a _test server_ rather than on a live/production server. Built new AI? One should have to test it on a test server, whether it’s a script or compiled code.

**Ed. Update**: More mitigations I'm thinking of now would be:
* Give adminstrators ability to boot the game w/out editor mode (without loading in-progress zones at all, and locking out builders until "all clear").
* If it becomes too difficult to ensure builders can't be malicious to the server: Provide eventing or other mechanisms for game system code to decide their own trust/restrictions for builder logins. Such as requiring the associated Player character to be a certain level (validated from game system code) and/or login hours (validated from Core code). This would help ensure they meet a minimum bar of understanding of the thematic nuances of this particular game world, and also have "more at stake" discouraging malicious behavior (as they could lose their Player character too, when caught).

---

> I'm not sure why you feel the need to have a builder:login for builders. Would you like to explain why you see the need to have a different login mechanism to have them exposed to building criteria? A simpler way to prevent them from messing with the 'live' world would be to set up roles that keep them from only working on areas that are not tagged as 'live'. Also many MUDs have different roles which require builders to interact with the world. While I'm not against breaking the mold of current MUD use, I think there are applications which are relevant to the current use.

Couple main reasons this is beneficial.
1) Separation of concerns: There are a large number of behaviors that need to differ based on whether a character is a builder. It is easiest to simply check a flag on the character at certain times, rather than using complicated checks against a large number of factors. Characters with "builder" permissions would be allowed to use commands such as "goto" (with restrictions that let them warp themselves to rooms that are within areas that they have builder permissions). Better not let them warp themselves out of combat, or out of a tricky situation surrounded by aggros, or lost in a maze, etc.  Fine, just add more restrictions? What about combat within builder zones? Better not give them any XP from mobs they are designing and "ensuring they work". Just detect that they finished combat while in a builder area to negate XP/reward gain, etc? Builder adds a modified uber weapon to their inventory? The abusive players WILL find ways to circumvent a complicated system. Maybe they find a way to poison their mob, disengage combat by warping to a non-builder zone just before the mobs death, etc. It's much easier to add simple checks against "is this character a builder right now" (like when assigning XP, "if this player is a builder, assign 0") than trying to cover all the myriad of potential abuses.
2) Community. It should be easy for a player to be logged in, building, still have an online presence within the world and thus getting recognition for the amount of work they're putting into building. The "who" commands could show that the player is in builder mode, etc. If they are forced to use a different ID or login to a different world or do build solely with a web-interface or offline, they lose this community presence.
3) Trust. Ties in to the comment on having different roles... If what you're talking about are live events and such, there are permissions systems in the plans to cover these scenarios. "Highly trusted" individuals needn't be the only ones allowed to make content, even live. Blocking them too aggressively is contrary to the advancement of the world. Validation and acceptance of built worlds, integrating them into the player world, and modifying any "published" part of the world, is another story entirely. These actions require trust.

> Is it really nessecary to have a custom built OLC? With being an integrated first-class Windows program, can't we easily expose some process to IIS and have an integrated Web Development Area Creator? OLC is an outdated concept that doesn't fit well in the concept of a new generation system. Also if you really want OLC, just do it ported to WM.

"Necessary" is not the term for such a feature. There are 3 building paradigms listed here. While it's possible to support all 3, that may be overkill, but it wouldn't necessarily be a bad thing.
(**Ed. Update**: Moving response to the #135 thread.)

In any case, ya this is just a proposed feature, and wouldn't be scheduled to come into acceptance/action until around v0.8.
Also, Universe Manager Lite is already under development.  And simply "porting" existing building systems wouldn't necessarily be easy either, as we have a much stronger approach to building our objects, with the extendable Behaviors system and such.  Most existing world editors wouldn't let you create interesting, extensible objects.

---

> I guess my question more is, why do we need to change the current paradigm where a builder position is more of an 'endgame' scenario where you also are not an active player?

Not a _need_. The current paradigm that I've seen with building is not that it's an "end-game" but more that there is a second server allotted to safe building, or more often for new MUDs, they seek builders who haven't even played the game yet. There is simply too much friction though for the most valuable potential contributors, being those who play and care about your game world, before actually get involved. I'm not suggesting we require every MUD based on this code base to adopt open-builder policies. Admins are free to strip the OLC features, or enable them only on a secondary "builder port" instance anyway, as they see fit, then require administrative approval to any builder accounts, then and add any other friction that they see fit, in order to enact the existing paradigms. I wouldn't do it that way, but as an open source basis for future MUDs, administrators should have options readily available.

> Valid points all. It's just my opionion that a quality built and integrated OLC is rarely worth the time and effort of development when a simple text file can be easily uploaded to a FTP port is the most powerful and quickest way to build the world.

Simple text file will work for basic room descriptions, exit mapping, and so on - but falls apart very quickly when you start talking about spawner rules, interesting interactivity, and so on. The format would have to be highly dependent on the game system when you start to talk about mobs and such.

There may be some compromise room too: Suppose we added support for some "common" file format that already exists for world building, to import as a "starter shell" of a real zone that the builder can then further iterate on.

> Ah. Right down the LPC/Diku split: I will also agree that new MUDs do seek builders who haven't played.

> So, we are talking about a paradigm shift in the way that Game Administrators and staff interact with the game and the world. While I'm not against it, I would suggest that as above by giving someone the ability to build we are revealing much about how the system works and removes a lot of the mystery of the game mechanics, limiting replayability.  Thus that character should be removed from active play.

Arguably a player can chose ruin their own experience in other ways (looking up this sort of information online, for example) regardless if we build this.  Should an administrator remove from play any player who they suspect has read the _open source code_ that their MUD is based on? Probably not, IMO. If they ruin the game experience for themselves, that's their own fault. Same with building, if reading up on the builder guides for proper stat allocations for a mob of a given level ruins some of the magic for them, and they still want to be a player, then they shouldn't do it. Further, a "remove from play" policy is completely unenforcable; nothing prevents them from having another character that they actually play.

> It is enforceable to a degree. Community standards, ipchecking and other methods can be used to enforce a one character per player ruleset. While there are methods to get around each of these it boils down to the culture of trust in the game.

Fair.

Anyway, personally, I've seen successful MUDs allow active players to do world building too, and there tends to be strong incentive for them to do so: Sometimes in-game rewards given to the builders. Even if not, there is always the inherent popularity feel-good factor, getting mad props from other players for having added a new zone to their game.  Balancing content creation rewards versus fairness can result in a healthy clip of content creation.

On the other hand: Having a "promotion to builder" model as incentive/reward for being a model helper-type-player has its own advantages too. Hence why administrators should be enabled to make these kinds of choices about which builder paradigms they want to support, instead of only having one choice available, or worse being requried to build the systems themselves from scratch.

> The complexity of a well built OLC that integrates well with a codebase is not worth the development costs. A system that require the MUD to interface with world development slows down development and takes away from allowing a RAD environment.

It will take some work to build, yes, but I do believe the benefits will be worth it.

I think OLC in WheelMUD is well poised to be a vertical slice that can be easily stripped, upgraded, and so on, independently of the rest of the MUD. The builder commands would be organized just like other Command files and isolated as such. Even during Rapid Application Development, where items and rooms and such are constantly gaining new Behaviors, the commands themselves needn't change: They will be aware of and operate with our Behaviors system in a _flexible and dynamic_ way. IE you'd make a Sword object, give it the Weapon behavior, which in and of itself has certain new Properties which are self-exposing to the builder system (via in-code Attributes if necessary), thus you'd be able to specify the damage (regardless of whether your game system uses "damage dice" or any other system) without having to change the core builder commands themselves. Then a builder tacks on the TriggersSpellBehavior (specified by name or whatnot, note that they can list applicable behaviors dynamically with another command) with properties specifying a fireball and percentage chance to proc, and now you've utilized advanced extension features that weren't availabe in the core WheelMUD code base, yet having never changed the OLC code. Can your standard desktop world-building tool make holy avenger swords which randomly shoot 30 die fireballs? Probably not, unless they are built with the same extensibility. By building our world-building tools in a way that is compatible with WheelMUD extensibility model, we certainly can make that sort of thing relatively painless and with no coding knowledge needed from the builder.

> I missused my words. As I mentioned above, the easiest way for me to code a new area is a simple text file that is uploaded via FTP. Learning and using an OLC heightens the misery of building for me. And in my opinion based on how I operate is a higher barrier to entry. Anyone remember the building staff in nightmare? That allowed you to change certain properties of a room/mob/item. It was horrible to use and fixing errors was extremly difficult. Even on an advanced OLC like the one found in DS 3.0 doesn't fix the problems of building an area with the ease of copying a room object text file and making the appropriate changes.

A tool like Universe Manager Lite is certainly going to be the best option for any who can go that route.  Probably most users would be able to go that route.  Some reasons one might not be able to though, might include: they don't run a supported OS, they can only contribute a little at a time, they want to collaborate on the area without having to pass files around, they aren't allowed/able to install programs where they want to build from (say on breaks at work or school or whatnot), etc.

For others, I strongly believe we can do _way better_ at user-friendly OLC than your nightmare example.

When we get near v0.8, we'll want to look at our world building options again, see how Universe Manager Lite is coming along, and decide if adding an OLC is worthwhile to do as well.

---

> I like the idea of people needing to create a test player account; Admins should be above account restrictions and it may be useful for an account to be used by an undercover admin who chooses to bump shoulders with the players directly.

---

> I like the idea of separate builder accounts.  It removes temptation to cheat while playing.  Other thoughts:
> 
> Firstly, you don't have to have it edit the production data.  Have it go to a development server and need to be approved to be moved into production.

Same with OLC or UML. I've described how content could need review by a higher level builder/admin before being integrated into the game via OLC, and UML doesn't need to operate against a production version of the DB either.

Yes, it would be nice if we can have solid "migration" features for content to help support those who want their world building server to be another isolated MUD instance - One key is the instances will need to maintain parity for "tested" content to still be valid (tested against the same code) when the migration occurs.  This sort of scenario will just take more thought/design to encourage admins to do well. :)

> If you wanted to be a massive massive MMO, then you would move to a model similar to Wikipedia.  Allow lots of builders in, with the ability to create links and such.  Probably limit them to just room descriptions or have a LOT of enforced rules.  Have a team of moderators that check on any submitted changes/additions.  Have the history tracking to be able to back-track.  I'm not saying we should do that as I think that's very niche, but it is a concept.

Interesting ideas.  Also will need more thought and specific design.

> Secondly, you say no community presence, I don't buy that either.  Online forums/chat?  You could also be sitting in the game as an admin while you have another window open to edit the world.

Fair: They player loses _some_ community presence, not _all_. (Unless they are using a strict single-window terminal like an old library system or super _retro_ scenario or something on purpose, for such extreme edge cases they'd lose all immediate presence.)  So usually, the player can _recover some of their missing community_ by juggling windows/tabs. Often this will sacrifice the _speed_ at which they can communicate, which lessens their community experience anyway from the "nearly face to face" feel of chatting/collaborating realtime in the same foreground window on the same server.

(**Ed. Update**: 3rd and 4th points are going to the #135 thread.)

> P.S. - I really like how you take constructive criticism and debate it.  A lot of people would get grumpy and give up.

Thanks!

---

> I still _personally_ don't see myself using OLC much except for GM stuff.  I prefer to build outside the game and just moderate inside the game.  However, I may need to get an item from the database and give it to someone in game.  I don't know if that's separate moderator (GM, etc.) interface or if you consider that being the same tool as the OLC.  All room/mob/object creating/editing/deleting, I would probably do offline.

Spawning an instance of a given thing in-game should be a single command, executable by admins with sufficient priveleges to do so.  The OLC designer style commands wouldn't really be needed for such cases, just the list commands to find the ID of the thing you want to spawn an instance of and the spawn command.
(In addition to it's other purposes, I'm not sure I touched on: OLC may be the fastest way for devs to test various combinations of Behaviors, Effects, their properties, and such, by creating and tweaking said combinations on a temporarily test Thing.  For world-building it's certainly not the ideal tool for everyone.)

---

> I think it is important to make this distinction. OLC CAN be used for world building but it WILL be used by the admins to create quest items and mobs, give new behaviours to things, etc. It may not be necessary to implement a full OLC but the difference between a simple OLC and a full OLC is minimal. I think that both systems should be made and both will be used which is the important thing.
